{"cells":[{"cell_type":"markdown","source":["[![Open in colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/nunososorio/bhs/edit/main/Data_structures/NB2_Intro_scanpy.ipynb)"],"metadata":{"id":"38fXOLWdo8Sh"},"id":"38fXOLWdo8Sh"},{"cell_type":"markdown","source":["# Notebook 2 - [AnnData](https://anndata.readthedocs.io/en/latest/) and [Scanpy](https://scanpy.readthedocs.io/en/stable/)\n","In this notebook we will explore the Annotated Data (AnnData) data structure and the [Scanpy](https://scanpy.readthedocs.io/en/stable/) library.\n","\n","This data structure is meant to allow you to keep (biological) data together with annotations about its observations (patients/cells/tumors...) and its variables"],"metadata":{"id":"0tZuREYp3uT2"},"id":"0tZuREYp3uT2"},{"cell_type":"markdown","id":"30886d4c-3b9a-4a06-a9d3-dd1d1877296c","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[],"id":"30886d4c-3b9a-4a06-a9d3-dd1d1877296c"},"source":["# Setup the environment\n","\n","The *basic* libraries (Numpy, Pandas, Matplotlib, Seaborn...) are already installed in Google Colab. To run this notebook you will need to install scanpy and anndata"]},{"cell_type":"code","source":["import sys\n","IN_COLAB = 'google.colab' in sys.modules\n","\n","if IN_COLAB:\n","    !pip install scanpy anndata"],"metadata":{"id":"xyn01C7C3UiW"},"id":"xyn01C7C3UiW","execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"id":"8f7f09ff-1e98-4b91-b237-2b78db8dc199","metadata":{"id":"8f7f09ff-1e98-4b91-b237-2b78db8dc199"},"outputs":[],"source":["# Import all the libraries we will use\n","import os\n","import numpy as np\n","from scipy.stats import ttest_ind\n","from scipy.stats import wilcoxon\n","import pandas as pd\n","import matplotlib.pyplot as plt\n","import seaborn as sns\n","import anndata\n","import random\n","import scanpy as sc"]},{"cell_type":"code","execution_count":null,"id":"bdc3bf57-9b4c-4bf9-9e27-3ef071e2f5f6","metadata":{"id":"bdc3bf57-9b4c-4bf9-9e27-3ef071e2f5f6"},"outputs":[],"source":["# Some details for the plots\n","plt.rcParams.update({'font.size':18, 'figure.figsize':(8,8)})"]},{"cell_type":"markdown","id":"9f6317d6-cd53-4f7e-9f29-ec638b5db49a","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[],"id":"9f6317d6-cd53-4f7e-9f29-ec638b5db49a"},"source":["# Load some data\n","\n","We will use the simulated data we generated previously. If you didn't sotre you can just get from the course repository.\n","\n"]},{"cell_type":"code","source":["if not os.path.exists('My_data.csv'):\n","    !wget https://raw.githubusercontent.com/Leo-GG/bhs/main/Data_structures/My_data.csv"],"metadata":{"id":"E6Ut2rjP2-I3"},"id":"E6Ut2rjP2-I3","execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"id":"099b5ac1-25ef-488f-a47c-96274582d2b3","metadata":{"id":"099b5ac1-25ef-488f-a47c-96274582d2b3"},"outputs":[],"source":["df=pd.read_csv('My_data.csv',index_col=0)"]},{"cell_type":"markdown","source":["If you remember, this data was a table where we had stored:\n","1. One row per observation\n","2. One column per \"gene\", with a numerical value for each observation\n","3. One column with an attribute of of the observations (diseased -D- or healthy -H-)"],"metadata":{"id":"nt2QDNPQ7DG8"},"id":"nt2QDNPQ7DG8"},{"cell_type":"code","execution_count":null,"id":"c1ca2aac-d0eb-4579-93b4-1aee965b8a2d","metadata":{"id":"c1ca2aac-d0eb-4579-93b4-1aee965b8a2d"},"outputs":[],"source":["# Have a look and see that it looks alright\n","df.head(5)"]},{"cell_type":"code","execution_count":null,"id":"a1c3206f-dbb5-437c-82ce-acf43b74b45a","metadata":{"id":"a1c3206f-dbb5-437c-82ce-acf43b74b45a"},"outputs":[],"source":["# Like before, we should have 100 observations and 81 columns/variables\n","df.shape"]},{"cell_type":"code","source":["# We saw that this is very useful, but here we have labels mixed up with values\n","\n","#df*df #This will give an error!"],"metadata":{"id":"mWs0hMopFNB2"},"id":"mWs0hMopFNB2","execution_count":null,"outputs":[]},{"cell_type":"code","source":["# We have to remember to separate labels from data!\n","df.iloc[:,:-1]*df.iloc[:,:-1]"],"metadata":{"id":"xl6rURvmFBwj"},"id":"xl6rURvmFBwj","execution_count":null,"outputs":[]},{"cell_type":"markdown","id":"b3d50283-8ff1-4fb4-a750-6518bd92c93c","metadata":{"tags":[],"id":"b3d50283-8ff1-4fb4-a750-6518bd92c93c"},"source":["# From DataFrame to AnnData\n","\n","The anndata library supports the use of Annotated Data (AnnData) objects. In these objects\n","- The numerical data is kept in a main matrix (X)\n","- Annotations about the observations and the variables are kept in accesory tables, names \"obs\" and \"var\"\n","<br/><br/>\n","<br/><br/>\n","\n","<img src=\"https://falexwolf.de/img/scanpy/anndata.svg\" alt=\"AnnData\" style=\"width:600px; height:auto;\"/>\n","\n","\n"]},{"cell_type":"markdown","id":"1bced1bf-c9c5-4e91-84dd-809f171cd88e","metadata":{"id":"1bced1bf-c9c5-4e91-84dd-809f171cd88e"},"source":["\n","<br/><br/>\n","We will build an AnnData object using our table:\n","- The \"gene\" data will go on the X matrix\n","- The information we have about the Condition will go to the \"obs\" DataFrame\n","- The \"gene\" names will go to the \"var\" DataFrame\n","\n","\n","\n","\n"]},{"cell_type":"code","execution_count":null,"id":"7412bc61-bed3-47b7-b9fc-68c4c6470795","metadata":{"id":"7412bc61-bed3-47b7-b9fc-68c4c6470795"},"outputs":[],"source":["import anndata\n","\n","# First we create the object using the table, keeping all the columns except the last one\n","adata=anndata.AnnData(X=df.iloc[:,:-1].copy())\n","adata\n"]},{"cell_type":"code","source":["# Then we copy the Condition to the \"obs\" DataFrame\n","adata.obs=df[['Condition']].copy()\n","adata"],"metadata":{"id":"2gmWsfCC97pe"},"id":"2gmWsfCC97pe","execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"id":"a6b8da0f-8b25-4b9e-9d50-cdcc24d0f02c","metadata":{"id":"a6b8da0f-8b25-4b9e-9d50-cdcc24d0f02c"},"outputs":[],"source":["# Let's read the warning and fix the issue...\n","adata.obs.index=adata.obs.index.astype(str)\n","adata"]},{"cell_type":"markdown","source":["## Let's check each element of our AnnData object"],"metadata":{"id":"4pAcpmU_3xaU"},"id":"4pAcpmU_3xaU"},{"cell_type":"code","execution_count":null,"id":"3632e262-d7d8-4168-a3e8-1b15b4ec149e","metadata":{"id":"3632e262-d7d8-4168-a3e8-1b15b4ec149e"},"outputs":[],"source":["adata.X"]},{"cell_type":"code","source":["adata.obs.head(3)"],"metadata":{"id":"aedhAV-c-CcB"},"id":"aedhAV-c-CcB","execution_count":null,"outputs":[]},{"cell_type":"code","source":["# The \"var\" DataFrame will be empty for now, but we can check that its index is made of \"gene\" names\n","adata.var.head(3)"],"metadata":{"id":"aHkidZSo9_g6"},"id":"aHkidZSo9_g6","execution_count":null,"outputs":[]},{"cell_type":"markdown","id":"d469f7dc-8689-4f08-9e8d-50416402f4e9","metadata":{"tags":[],"id":"d469f7dc-8689-4f08-9e8d-50416402f4e9"},"source":["# OK, but why?"]},{"cell_type":"markdown","id":"ee5938ce-1348-49fb-8a8a-0184c7011ca6","metadata":{"id":"ee5938ce-1348-49fb-8a8a-0184c7011ca6"},"source":["## 1. Efficient data storage\n","One very important feature of the AnnData object is that it separates numerical values from annotations\n","\n","The numerical data stored in the X matrix can now be stored using memory-efficient formats, such as sparse matrices\n","\n","This makes possible to store and access *very* large datasets"]},{"cell_type":"markdown","id":"522b2260-f12d-46c5-bd5a-b3067127c0f8","metadata":{"id":"522b2260-f12d-46c5-bd5a-b3067127c0f8"},"source":["## 2. Annotations\n","\n","Now we can add all the information we want about the observations/subjects/samples and about the features!\n","\n","Categorical and strings do not get mixed up with the numerical values on the matrix"]},{"cell_type":"code","execution_count":null,"id":"33d54a9f-716c-485f-b51a-cfa6589c3316","metadata":{"id":"33d54a9f-716c-485f-b51a-cfa6589c3316"},"outputs":[],"source":["# Maybe we want to include the age of our subjects\n","adata.obs['Age']=[ random.randint(20, 100) for i in range(adata.n_obs)]\n","\n","# Add their sex\n","adata.obs['Sex']=np.random.choice(['F','M'],adata.n_obs)\n","\n","# And some other variable\n","adata.obs['Hospitalized']=np.random.choice(['Y','N'],adata.n_obs)"]},{"cell_type":"code","execution_count":null,"id":"dd572efa-bbe6-4bb6-906f-40ac86f01c56","metadata":{"id":"dd572efa-bbe6-4bb6-906f-40ac86f01c56"},"outputs":[],"source":["adata.obs.head(3)"]},{"cell_type":"code","execution_count":null,"id":"3bf1c204-3c7f-4027-8381-40be07a4be58","metadata":{"id":"3bf1c204-3c7f-4027-8381-40be07a4be58"},"outputs":[],"source":["# Same with the genes, we can add some annotations to them\n","# Describe what the variable represents\n","adata.var['Type']='Not-a-real-gene'\n","\n","# Add a random clinical annotation\n","adata.var['Clinical_annotation']=np.random.choice(['Pathogenic','NA','Unknown','Developmental'],adata.n_vars)"]},{"cell_type":"code","execution_count":null,"id":"24689182-92c1-4d11-a6b4-f1990dc1ede4","metadata":{"id":"24689182-92c1-4d11-a6b4-f1990dc1ede4"},"outputs":[],"source":["adata.var.head(3)"]},{"cell_type":"code","execution_count":null,"id":"fffb02a9-143c-4b51-9035-b8b28f494eac","metadata":{"id":"fffb02a9-143c-4b51-9035-b8b28f494eac"},"outputs":[],"source":["# Now the AnnData object is storing all the information in a single object\n","adata"]},{"cell_type":"markdown","source":["\n","### We have now a rather complex object!\n","We have\n","- High-dimensional biological data\n","- Information about our observations\n","- Information about the variables\n","\n","All in one place!"],"metadata":{"id":"gpvzHj3e40IH"},"id":"gpvzHj3e40IH"},{"cell_type":"markdown","id":"984af53c-fa7e-4364-8cee-1f32903596bc","metadata":{"id":"984af53c-fa7e-4364-8cee-1f32903596bc"},"source":["## 3. Data analysis!\n","\n","The [scanpy](https://scanpy.readthedocs.io/en/stable/) library, designed for single-cell RNA-seq data analysis, offers *many* methods that can be used with data stored in AnnData objects\n","\n","<br/><br/>\n","\n","<img src=\"https://scanpy.readthedocs.io/en/stable/_static/Scanpy_Logo_BrightFG.svg\" alt=\"AnnData\" style=\"width:100px; height:auto;\"/>\n","\n","<br/><br/>\n"]},{"cell_type":"markdown","source":["### Plotting\n","Scanpy has great plotting functions!\n","You can plot your data referring to any of the annotations stored in \".var\" or in \".obs\""],"metadata":{"id":"9T8rqnc6AL1-"},"id":"9T8rqnc6AL1-"},{"cell_type":"code","source":["# For example, lets pick some genes to plot\n","adata.var.iloc[[3,8,22, 54, 12]]"],"metadata":{"id":"r7LIeXWRApHE"},"id":"r7LIeXWRApHE","execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"id":"3beb9276-ac0e-4eb6-a31b-97389766b3d6","metadata":{"id":"3beb9276-ac0e-4eb6-a31b-97389766b3d6"},"outputs":[],"source":["# Now let's use scanpy to plot their values on each of the two condition groups\n","\n","sc.pl.heatmap(adata=adata, var_names=['SOLB','EGFRE','MTE','PORT_NC','TACD'], groupby='Condition', swap_axes=True,figsize=[8,6])"]},{"cell_type":"code","source":["# It would be the same if we want to plot by another label\n","sc.pl.heatmap(adata=adata, var_names=['SOLB','EGFRE','MTE','PORT_NC','TACD'], groupby='Hospitalized', swap_axes=True,figsize=[8,6])"],"metadata":{"id":"DE-9xNYhHsGs"},"id":"DE-9xNYhHsGs","execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["### Statistical tests\n","Scanpy has functions to easily compare groups of observations\n"],"metadata":{"id":"ZtS8V4YcEmwm"},"id":"ZtS8V4YcEmwm"},{"cell_type":"code","execution_count":null,"id":"a58bab48-cc53-4db6-a7f3-3d76b3d7c6e4","metadata":{"id":"a58bab48-cc53-4db6-a7f3-3d76b3d7c6e4"},"outputs":[],"source":["# For example rank_genes_groups() allows you make comparisons between groups of observations\n","\n","sc.tl.rank_genes_groups(adata=adata, groupby='Condition', method='wilcoxon')"]},{"cell_type":"code","execution_count":null,"id":"c66eae3e-ead1-475b-9f05-3da8b6c27471","metadata":{"id":"c66eae3e-ead1-475b-9f05-3da8b6c27471"},"outputs":[],"source":["# The results are stored in the \"unstructured\" (uns) layer\n","\n","adata"]},{"cell_type":"code","source":["#adata.uns['rank_genes_groups']['pvals']"],"metadata":{"id":"ET315p9T59HN"},"id":"ET315p9T59HN","execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"id":"3a071ccb-9971-4e61-892f-975842b1cf4c","metadata":{"id":"3a071ccb-9971-4e61-892f-975842b1cf4c"},"outputs":[],"source":["# We can use these results to plot the features that are more different on each group\n","sc.pl.rank_genes_groups(adata)"]},{"cell_type":"code","execution_count":null,"id":"93a070f0-a41b-4e09-b8ce-61b7b1c90dde","metadata":{"id":"93a070f0-a41b-4e09-b8ce-61b7b1c90dde"},"outputs":[],"source":["# We can plot the actual values of the top-ranking variables - i.e. the ones with most differences across groups\n","\n","sc.pl.rank_genes_groups_heatmap(adata=adata,n_genes=20,groupby='Condition',swap_axes=True,figsize=[10,12])"]},{"cell_type":"code","execution_count":null,"id":"835e52cc-cd0b-4b65-8602-b17e6c8d458f","metadata":{"id":"835e52cc-cd0b-4b65-8602-b17e6c8d458f"},"outputs":[],"source":["# It also allows you to operate on your data, e.g. log-transforming it\n","\n","sc.pp.log1p(adata)\n"]},{"cell_type":"markdown","source":["### Feature selection\n","Scanpy allows you to **easily** do a simple feature selection:\n","the highly_variable_genes() function uses different methods to label variables as \"highly variable\" or not, based on their values across the dataset\n","<br/><br/>"],"metadata":{"id":"koR_Q3QcE4Qu"},"id":"koR_Q3QcE4Qu"},{"cell_type":"code","execution_count":null,"id":"a2f101ce-df29-4992-886b-717231684082","metadata":{"id":"a2f101ce-df29-4992-886b-717231684082"},"outputs":[],"source":["# Let's label our \"genes\"\n","\n","sc.pp.highly_variable_genes(adata=adata,min_disp=0)\n"]},{"cell_type":"code","source":["adata.var"],"metadata":{"id":"JyYUwS236pf8"},"id":"JyYUwS236pf8","execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"id":"39c99634-22fa-43f2-a86a-bb8f3487eb3b","metadata":{"id":"39c99634-22fa-43f2-a86a-bb8f3487eb3b"},"outputs":[],"source":["# You directly call a function to visualize the results of this function\n","\n","sc.pl.highly_variable_genes(adata_or_result=adata)"]},{"cell_type":"markdown","source":["### Dimensionality reduction\n","\n","Scanpy was designed to work with single-cell sequencing data. In this type of data, each observation is described by thousands of variables.\n","\n","A very useful/important step in the analysis of **highly-dimensional** data is dimensionality reduction. These are techniques that reflect the information across all variables in fewer dimensions.\n","\n","For example, this allows us to make 2d or 3d visualizations of our 80-gene data\n"],"metadata":{"id":"xHeoF96DDavW"},"id":"xHeoF96DDavW"},{"cell_type":"code","execution_count":null,"id":"0c038183-7ccc-4e07-a5aa-a9f1e31ad66d","metadata":{"id":"0c038183-7ccc-4e07-a5aa-a9f1e31ad66d"},"outputs":[],"source":["# You can apply PCA with a single function call\n","sc.pp.pca(data=adata, n_comps=20, use_highly_variable=True)\n","\n","# Results are stored in the varm and obsm layers\n","adata"]},{"cell_type":"code","execution_count":null,"id":"de8fa985-e663-4df7-9cf5-cee426f5d8de","metadata":{"id":"de8fa985-e663-4df7-9cf5-cee426f5d8de"},"outputs":[],"source":["# Visualize the PCA loadings\n","sc.pl.pca_variance_ratio(adata=adata, log=True, n_pcs=20)"]},{"cell_type":"code","execution_count":null,"id":"cc1fac82-994b-4e2d-ae4f-43d635e6f622","metadata":{"id":"cc1fac82-994b-4e2d-ae4f-43d635e6f622"},"outputs":[],"source":["# Visualize the data projected on the PCs, color by variable or observation attribute\n","sc.pl.pca(adata=adata, color=['SOLB','EGFRE','Condition'])\n","\n","#sc.pl.pca(adata=adata, color=['SOLB','EGFRE','Condition'], components=['4,5'])"]},{"cell_type":"markdown","id":"d39fb89a-ede4-4324-bf23-3a5cae4de497","metadata":{"id":"d39fb89a-ede4-4324-bf23-3a5cae4de497"},"source":["# You are not bound to Scanpy!\n","\n","Remember that the AnnData object is still \"just\" a collection of data entries: a central matrix connected to dataframes with annotations\n","\n","You can read, write and transform this data using any library with tools specific to your problem\n"]},{"cell_type":"code","execution_count":null,"id":"cc6fa451-d2ab-4299-93d6-1e9bccfa121d","metadata":{"id":"cc6fa451-d2ab-4299-93d6-1e9bccfa121d"},"outputs":[],"source":["# For example, let's use sklearn to cluster the data and write the results back the anndata object\n","\n","# First import a clustering tool from sklearn\n","from sklearn.cluster import KMeans\n","# Instantiate it\n","kmeans = KMeans(n_clusters=4, random_state=42)\n","\n","# Then we apply it to our data\n","adata.obs['kmeans_cluster'] = kmeans.fit_predict(adata.X)\n"]},{"cell_type":"code","source":["adata.obs"],"metadata":{"id":"39QQR448F-Bc"},"id":"39QQR448F-Bc","execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Make sure the results are kept as categories and not numbers, so the plots will be easier to understand\n","adata.obs['kmeans_cluster']=adata.obs['kmeans_cluster'].astype('category')"],"metadata":{"id":"3rHRwy7iGfSE"},"id":"3rHRwy7iGfSE","execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"id":"0d8dd59e-52a9-44a9-8d64-7f36260a237a","metadata":{"id":"0d8dd59e-52a9-44a9-8d64-7f36260a237a"},"outputs":[],"source":["# Now let's use again the scanpy function to visualize the results on the PCA space\n","sc.pl.pca(adata, color=['kmeans_cluster'])"]},{"cell_type":"markdown","source":["### Cluster on the PCA space instead...if we have time"],"metadata":{"id":"1f72ggkt7D5x"},"id":"1f72ggkt7D5x"},{"cell_type":"code","execution_count":null,"id":"e0d525b9-140f-48bb-8da3-258b3698201f","metadata":{"id":"e0d525b9-140f-48bb-8da3-258b3698201f"},"outputs":[],"source":["# Cluster the data again, but use the first 2 principal components only\n","adata.obsm['X_pca'].shape # The PCA data is stored here, it is n_obs x the number of PCs that we calculated"]},{"cell_type":"code","execution_count":null,"id":"01cf5672-7aad-45af-9b71-dca3139256f3","metadata":{"id":"01cf5672-7aad-45af-9b71-dca3139256f3"},"outputs":[],"source":["# The first 2 PCs are the first 2 columns\n","adata.obsm['X_pca'][:,0:2]\n","\n","# We can use k-means again\n","adata.obs['kmeans_PCA'] = kmeans.fit_predict(adata.obsm['X_pca'][:,0:2])"]},{"cell_type":"code","execution_count":null,"id":"afc95e71-804c-4852-a6f8-e753f4a2268d","metadata":{"id":"afc95e71-804c-4852-a6f8-e753f4a2268d"},"outputs":[],"source":["# See the results\n","sc.pl.pca(adata, color=['kmeans_PCA'])"]},{"cell_type":"markdown","source":["# Into the multi-ome: tables of tables\n","\n","Current *-omics* techniques tend more and more to be run in parallel, so we can have information on different **modalities** for the same subjects/cells/tissue...\n","\n","For example, you we can have RNA-sequencing data and proteomics data. In our example, we would have two data tables and one table with subject information/metadata.\n","\n","<img src=\"https://github.com/Leo-GG/bhs/blob/main/Data_structures/Illustrations/premise3.png?raw=true\" alt=\"AnnData\" style=\"width:100px; height:auto;\"/>\n","\n","<br/><br/>\n"],"metadata":{"id":"AaQSoztW7MTC"},"id":"AaQSoztW7MTC"},{"cell_type":"markdown","source":["**Luckily** for us, there is a Python library designed for multiomics data: [Muon](https://muon.readthedocs.io/en/latest/)\n","\n","This framework implements a Multiomics Data structure - **MuData**\n","\n","\n","<img src=\"https://github.com/scverse/muon/raw/master/docs/img/muon_header.png\" alt=\"AnnData\" style=\"width:100px; height:auto;\"/>\n","\n","<br/><br/>\n","\n","\n"],"metadata":{"id":"zKS8AgH08-8A"},"id":"zKS8AgH08-8A"},{"cell_type":"markdown","source":["In Muon we basically pile up multiple AnnData objects, one for each **modality**, and we keep sets of annotations that are common to the observations (subjects/cells..) in all modalities\n","\n","<img src=\"https://github.com/Leo-GG/bhs/blob/main/Data_structures/Illustrations/muon_paper.png?raw=true\" alt=\"AnnData\" style=\"width:100px; height:auto;\"/>\n","source: Bredikhin et al. \"MUON: multimodal omics analysis framework\", Genome Biology (2022)\n","<br/><br/>\n","\n","<img src=\"https://github.com/Leo-GG/bhs/blob/main/Data_structures/Illustrations/muon_diagram.png?raw=true\" alt=\"AnnData\" style=\"width:100px; height:auto;\"/>\n","<br/><br/>"],"metadata":{"id":"zW4hSMMh-cbY"},"id":"zW4hSMMh-cbY"},{"cell_type":"markdown","source":["The main advantage of the Muon framework is that allows the use of **multimodal integration** methods: [MOFA](https://biofam.github.io/MOFA2/)\n","\n","<br/><br/>\n","\n","\n","<img src=\"https://biofam.github.io/MOFA2/images/mofa_overview.png\" alt=\"AnnData\" style=\"width:100px; height:auto;\"/>\n","<br/><br/>\n","\n","But that's a topic for another time...\n","<br/><br/>"],"metadata":{"id":"14WXoVbKGyZy"},"id":"14WXoVbKGyZy"},{"cell_type":"code","source":[],"metadata":{"id":"-EeIQYpW-u2M"},"id":"-EeIQYpW-u2M","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.9"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}
